src/function/algebra/polynomialRoot.js

```diff
@@ -0,0 +1,152 @@
+   *     by the linear coefficient and subsequent coefficients of increasing powers.
+   * @return {Array} The distinct roots of the polynomial
+   */
+
+  return typed(name, {
+    'number|Complex, ...number|Complex': (constant, restCoeffs) => {
+      const coeffs = [constant, ...restCoeffs]
+      while (coeffs.length > 0 && isZero(coeffs[coeffs.length - 1])) {
+        coeffs.pop()
+      }
+      if (coeffs.length < 2) {
+        throw new RangeError(
+          `Polynomial [${constant}, ${restCoeffs}] must have a non-zero non-constant coefficient`)
+      }
+      switch (coeffs.length) {
+        case 2: // linear
+          return [unaryMinus(divide(coeffs[0], coeffs[1]))]
+        case 3: { // quadratic
+          const [c, b, a] = coeffs
+          const denom = multiply(2, a)
+          const d1 = multiply(b, b)
+          const d2 = multiply(4, a, c)
+          if (equalScalar(d1, d2)) return [divide(unaryMinus(b), denom)]
+          const discriminant = sqrt(subtract(d1, d2))
+          return [
+            divide(subtract(discriminant, b), denom),
+            divide(subtract(unaryMinus(discriminant), b), denom)
+          ]
+        }
+        case 4: { // cubic, cf. https://en.wikipedia.org/wiki/Cubic_equation
+          const [d, c, b, a] = coeffs
+          const denom = unaryMinus(multiply(3, a))
+          const D0_1 = multiply(b, b)
+          const D0_2 = multiply(3, a, c)
+          const D1_1 = add(multiply(2, b, b, b), multiply(27, a, a, d))
+          const D1_2 = multiply(9, a, b, c)
+          if (equalScalar(D0_1, D0_2) && equalScalar(D1_1, D1_2)) {
+            return [divide(b, denom)]
+          }
+          const Delta0 = subtract(D0_1, D0_2)
+          const Delta1 = subtract(D1_1, D1_2)
+          const discriminant1 = add(
+            multiply(18, a, b, c, d), multiply(b, b, c, c))
+          const discriminant2 = add(
+            multiply(4, b, b, b, d),
+            multiply(4, a, c, c, c),
+            multiply(27, a, a, d, d))
+          if (equalScalar(discriminant1, discriminant2)) {
+            return [
+              divide(
+                subtract(
+                  multiply(4, a, b, c),
+                  add(multiply(9, a, a, d), multiply(b, b, b))),
+                multiply(a, Delta0)), // simple root
+              divide(
+                subtract(multiply(9, a, d), multiply(b, c)),
+                multiply(2, Delta0)) // double root
+            ]
+          }
+          // OK, we have three distinct roots
+          let Ccubed
+          if (equalScalar(D0_1, D0_2)) {
+            Ccubed = Delta1
+          } else {
+            Ccubed = divide(
+              add(
+                Delta1,
+                sqrt(subtract(
+                  multiply(Delta1, Delta1), multiply(4, Delta0, Delta0, Delta0)))
+              ),
+              2)
+          }
+          const rawRoots = cbrt(Ccubed, !!'allRoots').toArray().map(
```

Comment:
What does `!!'allRoots'` mean here? I guess that can be replaced with `true`?

Reply:
Well, this is my personal bias that I think we have discussed that I think Boolean arguments should only very rarely be used for functions, because it is usually very difficult to have any idea what a literal 'true' or 'false' means as a function argument. Almost all Boolean arguments, in my experience, should be changed to two-element self-documenting enums. In this case, that would be {principalRoot, allRoots}, with the first one the default. So I am attempting to not be hypocritical and write self-documenting code: !!'allRoots' provides the correct value (indeed, `true`) to the existing function that was written to take a Boolean, but also clearly indicates to the reader of the code the semantics/purpose of that flag, which is to get  all of the cube roots returned. If you object of course I will make a commit changing to `true` but then I would love it if you would consider the virtue of in the long run eliminating such unreadable (i mean, unreadable in usage) Boolean flags.

Reply:
Ahh, of course. Yes I totally agree with out about using named arguments over unnamed arguments. 

I've regularly seen the following pattern to achieve the same:

```js
const rawRoots = cbrt(Ccubed, true /* allRoots */).toArray().map(
```

or

```js
const allRoots = true
const rawRoots = cbrt(Ccubed, allRoots).toArray().map(
```

That will not have any runtime impact. 

I have a slight preference for not using the "magic" `!!` with a string, it may confuse people like me 😜 . Would it be ok with you to use one of these other variants?

Reply:
Sure, added commit to switch to second variant. (I don't think with current JavaScript runtimes in recent browsers what I wrote has any runtime impact either, but I agree variant 2 avoids being slightly cryptic. Clearly best in the long run would be to gradually eliminate raw-boolean-flag arguments.)

Reply:
👍 thanks